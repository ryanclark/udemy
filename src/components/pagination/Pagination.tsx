import React from 'react';

import { RESULTS_PER_PAGE } from '../../services/characters/charactersService';
import styled from 'react-emotion';

interface PaginationProps {
  offset: number;
  total: number;
  onChange: (offset: number) => void;
}

const Container = styled('div')`
  display: flex;
  justify-content: center;
  padding-bottom: 50px;
`;

export const Button = styled('div')`
  color: #fff;
  background-color: #ec5252;
  padding: 11px 12px;
  font-size: 15px;
  border-radius: 6px;
  cursor: pointer;
  margin: 0 10px;
`;

export function Pagination(props: PaginationProps) {
  const links = [];

  // only show the back button if we've gone past the initial shown characters
  if (props.offset >= RESULTS_PER_PAGE) {
    links.push(
      <Button
        key="previous"
        onClick={() => props.onChange(props.offset - RESULTS_PER_PAGE)}
      >
        Previous
      </Button>
    );
  }

  // only show the next button if there are more results left
  if ((props.total - props.offset) > RESULTS_PER_PAGE) {
    links.push(
      <Button
        key="next"
        onClick={() => props.onChange(props.offset + RESULTS_PER_PAGE)}
      >
        Next
      </Button>
    );
  }

  return (
    <Container>
      { links }
    </Container>
  );
}
