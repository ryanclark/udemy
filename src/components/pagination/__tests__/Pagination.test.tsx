import React from 'react';

import { mount } from 'enzyme';

import { Button, Pagination } from '../Pagination';

describe('Pagination', () => {
  it('should render correctly', () => {
    const wrapper = mount(<Pagination offset={20} total={41} onChange={() => null} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render the next button when there are more results', () => {
    const wrapper = mount(<Pagination offset={20} total={41} onChange={() => null} />);

    expect(wrapper.find(Button).at(1).text()).toBe('Next');
  });

  it('should render the previous button when needed', () => {
    const wrapper = mount(<Pagination offset={20} total={41} onChange={() => null} />);

    expect(wrapper.find(Button).at(0).text()).toBe('Previous');
  });

  it('should only render the next button when on the first page', () => {
    const wrapper = mount(<Pagination offset={0} total={21} onChange={() => null} />);

    expect(wrapper.find(Button).length).toBe(1);
    expect(wrapper.find(Button).at(0).text()).toBe('Next');
  });

  it('should only render the previous button when on the last page', () => {
    const wrapper = mount(<Pagination offset={20} total={21} onChange={() => null} />);

    expect(wrapper.find(Button).length).toBe(1);
    expect(wrapper.find(Button).at(0).text()).toBe('Previous');
  });

  it('should render both buttons when not on the first or last page', () => {
    const wrapper = mount(<Pagination offset={20} total={41} onChange={() => null} />);

    expect(wrapper.find(Button).length).toBe(2);
    expect(wrapper.find(Button).at(0).text()).toBe('Previous');
    expect(wrapper.find(Button).at(1).text()).toBe('Next');
  });

  it('should trigger the handler when a button is clicked', () => {
    const onChange = jest.fn();

    const wrapper = mount(<Pagination offset={0} total={21} onChange={onChange} />);

    wrapper.find(Button).at(0).simulate('click');

    expect(onChange).toBeCalledWith(20);
  });
});
