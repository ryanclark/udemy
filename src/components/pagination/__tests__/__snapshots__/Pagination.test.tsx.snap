// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render correctly 1`] = `
<Pagination
  offset={20}
  onChange={[Function]}
  total={41}
>
  <Styled(div)>
    <div
      className="css-jm9tqh"
    >
      <Styled(div)
        key="previous"
        onClick={[Function]}
      >
        <div
          className="css-t42fz6"
          onClick={[Function]}
        >
          Previous
        </div>
      </Styled(div)>
      <Styled(div)
        key="next"
        onClick={[Function]}
      >
        <div
          className="css-t42fz6"
          onClick={[Function]}
        >
          Next
        </div>
      </Styled(div)>
    </div>
  </Styled(div)>
</Pagination>
`;
